sycl/test/reduction/reduction_usm.cpp

```diff
@@ -0,0 +1,87 @@
+// UNSUPPORTED: cuda
+// Reductions use work-group builtins not yet supported by CUDA.
+
+// UNSUPPORTED: linux
+// TODO: Enable the test for Linux when CI uses GPU driver 20.06.15619 or newer.
+
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// TODO: Enable the test for HOST when it supports intel::reduce() and barrier()
+
+// This test performs basic checks of parallel_for(nd_range, reduction, func)
+// with reductions initialized with USM var.
+
+#include "reduction_utils.hpp"
+#include <CL/sycl.hpp>
+#include <cassert>
+
+using namespace cl::sycl;
+
+template <typename T, int Dim, class BinaryOperation>
+class SomeClass;
+
+template <typename T, int Dim, class BinaryOperation>
+void test(T Identity, size_t WGSize, size_t NWItems) {
+  queue Q;
+  auto Dev = Q.get_device();
+  if (!Dev.get_info<info::device::usm_shared_allocations>())
+    return;
+
+  // Initialize.
+  T CorrectOut;
+  BinaryOperation BOp;
+
+  buffer<T, 1> InBuf(NWItems);
+  initInputData(InBuf, CorrectOut, Identity, BOp, NWItems);
+
+  T *ReduVarPtr = (T *)malloc_shared(sizeof(T), Dev, Q.get_context());
```

Comment:
I've realized that the best way to address my concern from https://github.com/intel/llvm/pull/1657#discussion_r422208619 is probably going to be to repeat this test for all of the different [types of USM pointer](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc#usm-allocations).  Reductions should work for memory allocated with `malloc_host`, `malloc_device` or `malloc_shared`.

Since we haven't decided yet what should happen with non-USM pointers, we should consider adding a [runtime check](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc#usm-allocations) to make sure that the user hasn't accidentally provided a non-USM pointer.

Reply:
@Pennycook , Your concern was regarding host memory, thus I added checks for USM pointers allocated via malloc_host() as well. Now alloc_shared/alloc_host memory is used/tested in reduction.
I don't see how reduction would work with var allocated via malloc_device(). That memory is assumed to be not accessible on HOST, right?

Reply:
> I don't see how reduction would work with var allocated via malloc_device(). That memory is assumed to be not accessible on HOST, right?

This is the case I'm worried about, and why I think there should be a test with `malloc_device`.

I appreciate that it may not be clear from the reduction proposal, but the intent was that USM reductions would behave the same way as buffer/accessor reductions.  The result of a reduction should not be written immediately back to the host, but into the device memory specified by the USM pointer or accessor.  If the result is needed on the device for a subsequent kernel, no "extra" host synchronization is required; if the result is needed on the host, it's the user's responsibility to get it there (e.g. by creating a host accessor, using an explicit `memcpy` from a device pointer, or synchronizing the queue before reading from a shared pointer).

Since the result isn't guaranteed to be written back to the host, simply copying between any temporary space used by the reduction (on the device) and the device pointer initially specified by the user should be sufficient.  The host has to initiate the copy, but doesn't ever need to read the value itself.  I don't know whether you can achieve that with `handler::copy`, or if you'd need to use something else, hence the question in https://github.com/intel/llvm/pull/1657#discussion_r422208619.

Does that make sense?

Reply:
@Pennycook , Thank you for the explanations. I hope I understood the idea and added check of USM memory allocated via malloc_device() to the LIT test.
Please see the 5th commit.

Reply:
Yes, this looks right to me.  Thanks, @v-klochkov!